extmod/modgzip.c

```diff
@@ -0,0 +1,450 @@
+        int header_wbits;
+        if (*wbits >= -15 && *wbits <= -8) {
+            // −8 to −15: Uses the absolute value of *wbits as the window size
+            // logarithm. The input must be a raw stream with no header or
+            // trailer.
+            *wbits = -*wbits;
+        } else if (*wbits >= 8 && *wbits <= 15) {
+            // +8 to +15: The base-two logarithm of the window size. The input
+            //  must include a zlib header and trailer.
+            if (uzlib_parse_zlib_gzip_header(&self->state[0].read.decomp, &header_wbits) != UZLIB_HEADER_ZLIB) {
+                // TODO how does cpython handle header_wbits != wbits?
+                mp_raise_OSError(MP_EINVAL);
+            }
+        } else if (*wbits >= 24 && *wbits <= 31) {
+            // +24 to +31 = 16 + (8 to 15): Uses the low 4 bits of the value
+            //  as the window size logarithm. The input must include a gzip
+            //  header and trailer.
+            *wbits = *wbits & 0xf;
+            if (uzlib_parse_zlib_gzip_header(&self->state[0].read.decomp, &header_wbits) != UZLIB_HEADER_GZIP) {
+                mp_raise_OSError(MP_EINVAL);
+            }
+        } else if (*wbits >= 40 && *wbits <= 47) {
+            // +40 to +47 = 32 + (8 to 15): Uses the low 4 bits of the value
+            //  as the window size logarithm, and automatically accepts
+            //  either the zlib or gzip format.
+            *wbits = *wbits & 0xf;
+            if (uzlib_parse_zlib_gzip_header(&self->state[0].read.decomp, &header_wbits) < 0) {
+                // TODO how does cpython handle header_wbits != wbits?
+                mp_raise_OSError(MP_EINVAL);
+            }
+        } else if (*wbits == 0) {
+            // 0: Automatically determine the window size from the zlib
+            // header. Only supported since zlib 1.2.3.5.
+            if (uzlib_parse_zlib_gzip_header(&self->state[0].read.decomp, wbits) != UZLIB_HEADER_ZLIB) {
+                mp_raise_OSError(MP_EINVAL);
+            }
+        } else {
+            mp_raise_ValueError(MP_ERROR_TEXT("wbits"));
+        }
+    #if MICROPY_PY_GZIP_COMPRESS
+    } else {
+        if (*wbits >= -15 && *wbits <= -9) {
+            // −9 to −15: Uses the absolute value of *wbits as the window size
+            // logarithm, while producing a raw output stream with no header or
+            // trailing checksum.
+            *wbits = -*wbits;
+        } else if (*wbits >= 9 && *wbits <= 15) {
+            // +9 to +15: The base-two logarithm of the window size, which
+            // therefore ranges between 512 and 32768. Larger values produce
+            // better compression at the expense of greater memory usage. The
+            // resulting output will include a zlib-specific header and trailer.
+            self->mode = GZIPFILE_MODE_WRITE_ZLIB;
+            self->state[0].write.input_checksum = 1; // ADLER32
+        } else if (*wbits >= 25 && *wbits <= 31) {
```

Comment:
I reckon it's worth allowing a smaller minimum wbits, to get a smaller history.  Eg 22?  Then the 9's about could be 6's.

Reply:
I think this is worthwhile as it's a significant memory saving for small inputs, even if it is another CPython diff.

I've implemented this in a separate commit, but it can be squashed later.